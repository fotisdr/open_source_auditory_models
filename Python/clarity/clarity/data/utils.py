"""Utilities for data generation."""

from __future__ import annotations

from pathlib import Path
from typing import Final, Literal

import numpy as np
import scipy
import scipy.io

SPEECH_FILTER: Final = np.array(
    scipy.io.loadmat(
        Path(__file__).parent / "params/speech_weight.mat",
        squeeze_me=True,
    )["filt"]
)


def better_ear_speechweighted_snr(target: np.ndarray, noise: np.ndarray) -> float:
    """Calculate effective better ear SNR.

    Args:
        target (np.ndarray):
        noise (np.ndarray):

    Returns:
        (float)
    Maximum Signal Noise Ratio between left and right channel.
    """
    if np.ndim(target) == 1:
        # analysis left ear and right ear for single channel target
        left_snr = speechweighted_snr(target, noise[:, 0])
        right_snr = speechweighted_snr(target, noise[:, 1])
    else:
        # analysis left ear and right ear for two channel target
        left_snr = speechweighted_snr(target[:, 0], noise[:, 0])
        right_snr = speechweighted_snr(target[:, 1], noise[:, 1])
    # snr is the max of left and right
    be_snr = max(left_snr, right_snr)
    return be_snr


def speechweighted_snr(target: np.ndarray, noise: np.ndarray) -> float:
    """Apply speech weighting filter to signals and get SNR.

    Args:
        target (np.ndarray):
        noise (np.ndarray):

    Returns:
        (float):
    Signal Noise Ratio
    """

    target_filt = scipy.signal.convolve(
        target, SPEECH_FILTER, mode="full", method="fft"
    )
    noise_filt = scipy.signal.convolve(noise, SPEECH_FILTER, mode="full", method="fft")

    # rms of the target after speech weighted filter
    targ_rms = np.sqrt(np.mean(target_filt**2))

    # rms of the noise after speech weighted filter
    noise_rms = np.sqrt(np.mean(noise_filt**2))
    sw_snr = np.divide(targ_rms, noise_rms)
    return sw_snr


def sum_signals(signals: list) -> np.ndarray | Literal[0]:
    """Return sum of a list of signals.

    Signals are stored as a list of ndarrays whose size can vary in the first
    dimension, i.e., so can sum mono or stereo signals etc.
    Shorter signals are zero padded to the length of the longest.

    Args:
        signals (list): List of signals stored as ndarrays

    Returns:
        np.ndarray: The sum of the signals

    """
    max_length = max(x.shape[0] for x in signals)
    return sum(pad(x, max_length) for x in signals)


def pad(signal: np.ndarray, length: int) -> np.ndarray:
    """Zero pad signal to required length.

    Assumes required length is not less than input length.

    Args:
        signal (np.array):
        length (int):

    Returns:
        np.array:
    """

    if length < signal.shape[0]:
        raise ValueError("Length must be greater than signal length")

    return np.pad(
        signal, ([(0, length - signal.shape[0])] + ([(0, 0)] * (len(signal.shape) - 1)))
    )
