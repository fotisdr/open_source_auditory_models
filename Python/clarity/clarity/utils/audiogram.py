"""Dataclass to represent a monaural audiogram"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import TYPE_CHECKING, Callable, Final

import numpy as np

if TYPE_CHECKING:
    from numpy import ndarray

DEFAULT_CLARITY_AUDIOGRAM_FREQUENCIES: Final = np.array(
    [
        250,
        500,
        1000,
        2000,
        3000,
        4000,
        6000,
        8000,
    ]
)

FULL_STANDARD_AUDIOGRAM_FREQUENCIES: Final = np.array(
    [
        125,
        250,
        500,
        750,
        1000,
        1500,
        2000,
        3000,
        4000,
        6000,
        8000,
        10000,
        12000,
        14000,
        16000,
    ]
)


@dataclass
class Audiogram:
    """Dataclass to represent an audiogram.

    Attributes:
        levels (ndarray): The audiometric levels in dB HL
        frequencies (ndarray): The frequencies at which the levels are measured

    """

    levels: np.ndarray
    frequencies: np.ndarray = field(
        default_factory=lambda: np.array(DEFAULT_CLARITY_AUDIOGRAM_FREQUENCIES)
    )

    def __post_init__(self) -> None:
        """Check that dimensions of levels and frequencies match."""

        # Ensure that levels and frequencies are numpy arrays
        self.levels = np.array(self.levels)
        self.frequencies = np.array(self.frequencies)

        if len(self.levels) != len(self.frequencies):
            raise ValueError(
                f"Levels ({len(self.levels)}) and frequencies ({len(self.frequencies)})"
                " must have the same length"
            )

        if len(self.frequencies) != len(np.unique(self.frequencies)):
            raise ValueError("Frequencies must be unique")

        if not np.all(np.diff(self.frequencies) > 0):
            raise ValueError("Frequencies must be in ascending order")

    @property
    def severity(self) -> str:
        """Categorise HL severity level for the audiogram.

        Note that this categorisation is different from that of the British
        Society of Audiology, which recommends descriptors mild, moderate,
        severe and profound for average hearing threshold levels at 250, 500,
        1000, 2000 and 4000 Hz of 21-40 dB HL, 41-70 dB HL, 71-95 dB HL
        and > 95 dB HL, respectively (BSA Pure-tone air-conduction and
        bone-conduction threshold audiometry with and without masking
        2018).

        Returns:
            str -- severity level, one of SEVERE, MODERATE, MILD, NOTHING

        """
        # calculate mean hearing loss between critical frequencies of 2 & 8 kHz
        critical_freqs = np.logical_and(
            2000 <= self.frequencies, self.frequencies <= 8000
        )
        critical_levels = self.levels[critical_freqs]
        # Remove any None values
        critical_levels = [x for x in critical_levels if x is not None]
        # Ignore any None values
        impairment_degree = np.mean(critical_levels) if len(critical_levels) > 0 else 0

        if impairment_degree > 56.0:
            return "SEVERE"
        if impairment_degree > 35.0:
            return "MODERATE"
        if impairment_degree > 15.0:
            return "MILD"

        return "NOTHING"

    def has_frequencies(self, frequencies: ndarray) -> bool:
        """Check if the audiogram has the given frequencies.

        Args:
            frequencies (ndarray): The frequencies to check

        Returns:
            bool: True if the audiogram has the given frequencies

        """
        return np.all(np.isin(frequencies, self.frequencies, assume_unique=True))

    def resample(
        self, new_frequencies: ndarray, linear_frequency: bool = False
    ) -> Audiogram:
        """Resample the audiogram to a new set of frequencies.

        Interpolates linearly on a (by default) log frequency axis. If
        linear_frequencies is set True then interpolation is done on a linear
        frequency axis.

        Args:
            new_frequencies (ndarray): The new frequencies to resample to

        Returns:
            Audiogram: New audiogram with resampled frequencies

        """

        # Either log frequency scaling or linear frequency scaling
        axis_fn: Callable = (lambda x: x) if linear_frequency else np.log

        return Audiogram(
            levels=np.interp(
                axis_fn(new_frequencies),
                axis_fn(self.frequencies),
                self.levels,
                left=self.levels[0],
                right=self.levels[-1],
            ),
            frequencies=new_frequencies,
        )


# Reference audiograms originally defined in the Cambridge group
# MSBG model MATLAB code.

# No loss
AUDIOGRAM_REF: Final = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
)

# No loss
AUDIOGRAM_REF_CLARITY: Final = Audiogram(
    frequencies=DEFAULT_CLARITY_AUDIOGRAM_FREQUENCIES,
    levels=np.array([0, 0, 0, 0, 0, 0, 0, 0]),
)

# Mild age-related hearing loss, slightly reduced from CF, first-time aid
# wearers group (used in Stafa talk by MAS).
AUDIOGRAM_MILD: Final = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([5, 10, 15, 18, 19, 22, 25, 28, 31, 35, 38, 40, 40, 45, 50]),
)

# Moderate hearing loss based on mild N2 flat/mod sloping
# from Bisgaard et al. 2020
AUDIOGRAM_MODERATE: Final = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([15, 20, 20, 22.5, 25, 30, 35, 40, 45, 50, 55, 55, 60, 65, 65]),
)

# Moderate-severe age-related hearing loss, average of MAS/KA summer proj 2011,
# elderly HI
AUDIOGRAM_MODERATE_SEVERE: Final = Audiogram(
    frequencies=FULL_STANDARD_AUDIOGRAM_FREQUENCIES,
    levels=np.array([19, 19, 28, 35, 40, 47, 52, 56, 58, 58, 63, 70, 75, 80, 80]),
)


@dataclass
class Listener:
    """Dataclass to represent a Listener.

    The listener is currently defined by their left and right ear
    audiogram. In later versions, this may be extended to include
    further audiometric data.

    The class provides methods for reading metadata files which
    will also include some basic validation.

    Attributes:
        id (str): The ID of the listener
        audiogram_left (Audiogram): The audiogram for the left ear
        audiogram_right (Audiogram): The audiogram for the right ear
    """

    audiogram_left: Audiogram
    audiogram_right: Audiogram
    id: str = ""

    @staticmethod
    def from_dict(listener_dict: dict) -> Listener:
        """Create a Listener from a dict.

        The dict structure and fields are based on those used
        in the Clarity metadata files.

        Args:
            listener_dict (dict): The listener dict

        Returns:
            Listener: The listener

        """
        return Listener(
            id=listener_dict["name"],
            audiogram_left=Audiogram(
                levels=listener_dict["audiogram_levels_l"],
                frequencies=listener_dict["audiogram_cfs"],
            ),
            audiogram_right=Audiogram(
                levels=listener_dict["audiogram_levels_r"],
                frequencies=listener_dict["audiogram_cfs"],
            ),
        )

    @staticmethod
    def load_listener_dict(filename: Path) -> dict[str, Listener]:
        """Read a Clarity Listener dict file.

        The standard Clarity metadata files presents listeners as a
        dictionary of listeners, keyed by listener ID.

        Args:
            filename (Path): The path to the listener dict file

        Returns:
            dict[str, Listener]: A dict of listeners keyed by id

        """
        with open(filename, encoding="utf-8") as fp:
            listeners_raw = json.load(fp)

        listeners = {}
        for listener_id, listener_dict in listeners_raw.items():
            listeners[listener_id] = Listener.from_dict(listener_dict)
        return listeners
