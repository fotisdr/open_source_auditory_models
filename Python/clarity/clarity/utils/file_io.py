"""File I/O functions."""

from __future__ import annotations

import json
import logging
from pathlib import Path

import numpy as np
import soundfile
from numpy import ndarray
from soundfile import SoundFile

from clarity.utils.signal_processing import resample

# Function for reading and writing jsonl files


def read_jsonl(filename: str) -> list:
    """Read a jsonl file into a list of dictionaries."""
    with open(filename, encoding="utf-8") as fp:
        records = [json.loads(line) for line in fp]
    return records


def write_jsonl(filename: str, records: list) -> None:
    """Write a list of dictionaries to a jsonl file."""
    with open(filename, "a", encoding="utf-8") as fp:
        for record in records:
            fp.write(json.dumps(record) + "\n")


# Functions for reading and writing audio signals.
#
# - Signals are passed and returns as numpy arrays of floats of shape
#   (n_samples, n_channels). Mono signals can be passed as (n_samples, 1)
#   but are returned as (n_samples,) which is more convenient for most purposes.
#   Floating point and 16 bit integer file formats are supported.
# - Signals are assumed to be in the range [-1.0 to 1.0) and will be scaled to
#   and from -32768 to 32767 when writing to or reading from 16 bit files.
#   On writing to 16 bit files, overflow can be set to silently wrap or to raise
#   an error.
# - Any sample rate can be used for writing and floating points values are
#   rounded to integers. For reading, the 'expected sample rate' is supplied and
#   the file is resample if necessary or can be set to throw an error. Again,
#   this is more convenient for most purposes.


def write_signal(
    filename: str | Path,
    signal: ndarray,
    sample_rate: float,
    floating_point: bool = True,
    strict: bool = False,
) -> None:
    """Write a signal as fixed or floating point wav file.

    Signals are passed as numpy arrays of floats of shape (n_samples, n_channels)
    for n_channels >= 1 or (n_samples,) for n_channels = 1.

    Signals are floating point in the range [-1.0 to 1.0) but can be written
    as wav file with either 16 bit integers or floating point. In the former,
    the signals are scaled to map to the range -32768 to 32767 and clipped
    if necessary.

    NB: setting 'strict' to True will raise error on overflow. This would be
    a more natural default but it would break existing code that did not
    check for overflow.

    Args:
        filename (str|Path): name of file in to write to.
        signal (ndarray): signal to write.
        sample_rate (float): sampling frequency.
        floating_point (bool): write as floating point else an ints (default: True).
        strict (bool): raise error if signal out of range for int16 (default: False).
    """

    if floating_point is False:
        subtype = "PCM_16"
        # Signal is float and we want to convert to int16
        # *NB* Not  *= in next line as we need to make a copy
        signal = signal * 32768.0
        if np.max(signal) > 32767.0 or np.min(signal) < -32768.0:
            if strict:
                raise ValueError("Signal out of range [-1.0, 1.0)")
            logging.warning(
                f"Writing {filename}. Signal out of range [-1.0, 1.0) - clipping."
            )
            signal = np.clip(signal, -32768.0, 32767.0)
        signal = signal.astype(np.dtype("int16"))
    else:
        subtype = "FLOAT"

    soundfile.write(filename, signal, sample_rate, subtype=subtype)


def read_signal(
    filename: str | Path,
    sample_rate: float = 0.0,
    offset: int | float = 0,  # offset can be in samples or seconds
    n_samples: int = -1,
    n_channels: int = 0,
    offset_is_samples: bool = False,
    allow_resample: bool = True,
) -> ndarray:
    """Read a wav format audio file and return as numpy array of floats.

    The returned value will be a numpy array of floats of shape (n_samples,
    n_channels) for n_channels >= 2, and shape (n_samples,) for n_channels = 1.

    If n_samples is set to a value other than -1, the specified number of samples
    will be read, or until the end of the file. An 'offset' can be set to start
    reading from a specified sample or time (in seconds).

    The expected number of channels can be specified. If the file has a different
    number of channels, an error will be raised.

    The expected sample rate can be specified. If the file has a different sample
    the file will be resampled to the expected sample rate, unless
    'allow_resample' is set to False, in which case an error will be raised.

    Args:
        filename (str|Path): Name of file to read
        sample_rate (float): The expected sample rate (default: 0.0 = any rate OK)
        offset (int, optional): Offset in samples or seconds (from start).
            Default is 0.
        n_samples (int): Number of samples.
        n_channels (int): expected number of channel (default: 0 = any number OK)
        offset_is_samples (bool): is offset measured in samples, (True) or
            seconds (False) (default: False)
        allow_resample (bool): allow resampling if sample rate is different
            from expected rate. Else raise error (default: True)

    Returns:
        np.ndarray: audio signal
    """

    wave_file = SoundFile(filename)

    if n_channels not in (0, wave_file.channels):
        raise ValueError(
            f"Wav file ({filename}) was expected to have {n_channels} channels."
        )

    if not offset_is_samples:  # Default behaviour
        offset = np.rint(offset * wave_file.samplerate).astype(int)

    if offset != 0:
        wave_file.seek(offset)

    signal = wave_file.read(frames=n_samples)

    if sample_rate not in (0, wave_file.samplerate):
        if allow_resample:
            signal = resample(signal, wave_file.samplerate, sample_rate)
        else:
            raise ValueError(
                f"Sample rate of {wave_file.samplerate} "
                "does not match expected {sample_rate}"
            )

    return signal
