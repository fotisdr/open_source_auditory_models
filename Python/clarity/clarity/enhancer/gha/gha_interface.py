from __future__ import annotations

import logging
import os
import pathlib
import subprocess
import tempfile
from pathlib import Path

import numpy as np
from jinja2 import Environment, FileSystemLoader

from clarity.enhancer.gha.gha_utils import format_gaintable, get_gaintable
from clarity.utils.audiogram import Listener
from clarity.utils.file_io import read_signal, write_signal


class GHAHearingAid:
    def __init__(
        self,
        sample_rate=44100,
        ahr=20,
        audf=None,
        cfg_file="prerelease_combination4_smooth",
        noise_gate_levels=None,
        noise_gate_slope=0,
        cr_level=0,
        max_output_level=100,
        equiv_0db_spl=100,
        test_nbits=16,
    ):
        if audf is None:
            audf = [250, 500, 1000, 2000, 3000, 4000, 6000, 8000]
        if noise_gate_levels is None:
            noise_gate_levels = [38, 38, 36, 37, 32, 26, 23, 22, 8]

        self.sample_rate = sample_rate
        self.ahr = ahr
        self.audf = audf
        self.cfg_file = cfg_file
        self.noise_gate_levels = noise_gate_levels
        self.noise_gate_slope = noise_gate_slope
        self.cr_level = cr_level
        self.max_output_level = max_output_level
        self.equiv_0db_spl = equiv_0db_spl
        self.test_nbits = test_nbits

    def create_configured_cfgfile(
        self, input_file, output_file, formatted_sGt, cfg_template_file
    ):
        """Using Jinja2, generates cfg file for given configuration.

        Creates template output file and configures with correct filenames, peak level
        out and DC gaintable.

        Args:
            input_file (str): file to process
            output_file (str): file in which to store processed file
            formatted_sGt (ndarray): gaintable formatted for input into cfg file
            cfg_template_file: configuration file template
            ahr (int): amplification headroom

        Returns:
            cfg_filename (str): cfg filename
        """

        if self.sample_rate != 44100:
            logging.error("Current GHA configuration requires 44.1kHz sampling rate.")
            raise ValueError(
                "Current GHA configuration requires 44.1kHz sampling rate."
            )

        cfg_template_file = pathlib.Path(cfg_template_file)

        # Define cfg filenames
        # Read new file and replace any parameter values necessary
        # Update peaklevel out by adding headroom
        logging.info("Adding %s dB headroom", self.ahr)

        peaklevel_in = int(self.equiv_0db_spl)
        peaklevel_out = int(self.equiv_0db_spl + self.ahr)

        # Render jinja2 template
        file_loader = FileSystemLoader(cfg_template_file.parent)
        env = Environment(loader=file_loader)
        template = env.get_template(cfg_template_file.name)
        output = template.render(
            io_in=input_file,
            io_out=output_file,
            peaklevel_in=(
                f"[{peaklevel_in} {peaklevel_in} {peaklevel_in} {peaklevel_in}]"
            ),
            peaklevel_out=f"[{peaklevel_out} {peaklevel_out}]",
            gtdata=formatted_sGt,
        )

        return output

    def process_files(
        self, infile_names: list[str], outfile_name: str, listener: Listener
    ):
        """Process a set of input signals and generate an output.

        Args:
            infile_names (list[str]): List of input wav files. One stereo wav
                file for each hearing device channel
            outfile_name (str): File in which to store output wav files
            dry_run (bool): perform dry run only
        """
        logging.info("Processing %s with listener %s", outfile_name, listener.id)

        logging.info(
            "Audiogram severity is %s (left) and %s (right)",
            listener.audiogram_left.severity,
            listener.audiogram_right.severity,
        )
        audiogram_left = listener.audiogram_left.resample(self.audf)
        audiogram_right = listener.audiogram_right.resample(self.audf)

        # Get gain table with noisegate correction
        gaintable = get_gaintable(
            audiogram_left,
            audiogram_right,
            self.noise_gate_levels,
            self.noise_gate_slope,
            self.cr_level,
            self.max_output_level,
        )
        formatted_sGt = format_gaintable(gaintable, noisegate_corr=True)

        cfg_template = Path(__file__).parent / f"cfg_files/{self.cfg_file}_template.cfg"

        # Merge CH1 and CH3 files. This is the baseline configuration.
        # CH2 is ignored.
        fd_merged, merged_filename = tempfile.mkstemp(
            prefix="clarity-merged-", suffix=".wav"
        )
        # Only need file name; must immediately close the unused file handle.
        os.close(fd_merged)

        self.create_HA_inputs(infile_names, merged_filename)

        # Create the openMHA config file from the template
        fd_cfg, cfg_filename = tempfile.mkstemp(
            prefix="clarity-openmha-", suffix=".cfg"
        )
        # Again, only need file name; must immediately close the unused file handle.
        os.close(fd_cfg)
        with open(cfg_filename, "w", encoding="utf-8") as f:
            f.write(
                self.create_configured_cfgfile(
                    merged_filename, outfile_name, formatted_sGt, cfg_template
                )
            )

        # Process file using configured cfg file
        # Suppressing OpenMHA output with -q - comment out when testing
        # Append log of OpenMHA commands to /cfg_files/logfile
        subprocess.run(
            [
                "mha",
                "-q",
                "--log=logfile.txt",
                f"?read:{cfg_filename}",
                "cmd=start",
                "cmd=stop",
                "cmd=quit",
            ],
            check=True,
        )

        # Delete temporary files.
        os.remove(merged_filename)
        os.remove(cfg_filename)

        # Check output signal has energy in every channel
        sig = read_signal(
            outfile_name, sample_rate=self.sample_rate, allow_resample=False
        )

        if len(np.shape(sig)) == 1:
            sig = np.expand_dims(sig, axis=1)

        if not np.all(np.sum(abs(sig), axis=0)):
            raise ValueError("Channel empty.")

        # Rewriting as floating point
        write_signal(outfile_name, sig, self.sample_rate, floating_point=True)

        logging.info("OpenMHA processing complete")

    def create_HA_inputs(self, infile_names: list[str], merged_filename: str) -> None:
        """Create input signal for baseline hearing aids.

        The baseline hearing aid takes a 4-channel wav file as input. This is
        constructed from the left and right signals of the front (CH1) and
        rear (CH3) microphones that are available in the Clarity data.

          Args:
              infile_names (list[str]): Names of file to read
              merged_file_name (str): Name of file to write

          Raises:
              ValueError: If input channel names are inconsistent
        """
        if (infile_names[0][-5] != "1") or (infile_names[2][-5] != "3"):
            raise ValueError("HA-input signal error: channel mismatch!")

        signal_CH1 = read_signal(
            infile_names[0], sample_rate=self.sample_rate, allow_resample=False
        )
        signal_CH3 = read_signal(
            infile_names[2], sample_rate=self.sample_rate, allow_resample=False
        )

        merged_signal = np.zeros((len(signal_CH1), 4))

        # channel index 0 = front microphone on the left hearing aid
        merged_signal[:, 0] = signal_CH1[:, 0]
        # channel index 1 = front microphone on the right hearing aid
        merged_signal[:, 1] = signal_CH1[:, 1]
        # channel index 2 = rear microphone on the left hearing aid
        merged_signal[:, 2] = signal_CH3[:, 0]
        # channel index 3 = rear microphone on the right hearing aid
        merged_signal[:, 3] = signal_CH3[:, 1]

        write_signal(
            merged_filename,
            merged_signal,
            self.sample_rate,
            floating_point=True,
            strict=True,
        )
