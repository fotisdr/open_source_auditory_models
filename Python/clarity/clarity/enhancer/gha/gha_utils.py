from __future__ import annotations

import logging
import re
from typing import TYPE_CHECKING, TypedDict

import numpy as np
from scipy.interpolate import interp1d

from clarity.enhancer.gha.gainrule_camfit import gainrule_camfit_compr

if TYPE_CHECKING:
    from numpy import ndarray

    from clarity.utils.audiogram import Audiogram


class Gaintable(TypedDict):
    """Gaintable for a given audiogram."""

    sGt_uncorr: ndarray
    sGt: ndarray
    noisegatelevel: ndarray
    noisegateslope: ndarray
    frequencies: ndarray
    levels: ndarray
    channels: int


class FittingParams(TypedDict):
    """Fitting parameters for gaintable calculation."""

    frequencies: ndarray
    edge_frequencies: ndarray
    levels: ndarray
    channels: int
    side: str


def get_gaintable(
    audiogram_left: Audiogram,
    audiogram_right: Audiogram,
    noisegate_levels: float | ndarray,
    noisegate_slope: float | ndarray,
    cr_level: float,
    max_output_level: float,
) -> Gaintable:
    """Compute a gaintable for a given audiogram.

    Replaces MATLAB GUI interface of original OpenMHA software for
    gaintable_camfit_compr table calculation. Assumes two channels and
    that audiogram frequencies are identical at two ears.

    Args:
        audiogram_left (Audiogram): the audiogram for the left ear
        audiogram_right (Audiogram): the audiogram for the right ear
        audf (list): audiogram frequencies for fitting
        noisegatelevels (ndarray): compression threshold levels for each frequency band
        noisegateslope (float): determines slope of gains below compression threshold
        cr_level (int): overall input level in dB for calculation of compression ratios
        max_output_level (int): maximum output level in dB

    Returns:
        dict: dim ndarray of gains

    """

    # Fixed centre frequencies for amplification bands
    sFitmodel: FittingParams = {
        "frequencies": np.array(
            [177, 297, 500, 841, 1414, 2378, 4.0000e03, 6.7270e03, 11314]
        ),
        "edge_frequencies": np.array(
            [
                1.0000e-08,
                229.2793,
                385.3570,
                648.4597,
                1.0905e03,
                1.8337e03,
                3.0842e03,
                5.1873e03,
                8.7241e03,
                10000001,
            ]
        ),
        "levels": np.arange(-10, 110 + 1),  # Levels SPL at which to compute the gains
        "channels": 2,
        "side": "lr",
    }

    # Calculate gains and compression ratios
    gain_table, noisegate_levels, noisegate_slope = gainrule_camfit_compr(
        audiogram_left,
        audiogram_right,
        sFitmodel,
        noisegate_levels,
        noisegate_slope,
        cr_level,
        max_output_level,
    )

    gain_table_corrected = multifit_apply_noisegate(
        gain_table,
        sFitmodel["frequencies"],
        sFitmodel["levels"],
        noisegate_levels,
        noisegate_slope,
    )

    # Reshape sGt here to suit cfg file input
    sGt_data: Gaintable = {
        "sGt_uncorr": gain_table,  # sGt without noisegate correction
        "sGt": np.transpose(np.reshape(gain_table_corrected, (121, 18), order="F")),
        "noisegatelevel": noisegate_levels,
        "noisegateslope": noisegate_slope,
        "frequencies": sFitmodel["frequencies"],
        "levels": sFitmodel["levels"],
        "channels": sFitmodel["channels"],
    }
    return sGt_data


def format_gaintable(
    gaintable: Gaintable,
    noisegate_corr: bool = True,
) -> str:
    """
    Format gaintable for insertion into cfg file as long string.

    Args:
        gaintable (ndarray): The gaintable to format
        noisegate_corr (boolean, optional): apply noisegate correction or do not
            (default: True)

    Returns:
        str: gaintable formatted for insertion into OpenMHA
            cfg file

    """
    if noisegate_corr:
        sGt = gaintable["sGt"]
    else:
        logging.warning("Noise gate correction not being applied to gain table.")
        sGt = gaintable["sGt_uncorr"]

    # Do not apply gains to 9th and 18th row
    logging.info("No application of gains to 9th and 18th row of table.")
    sGt[[8, -1], :] = 0

    # Format for inclusion in cfg file
    v = "["
    for k in range(0, np.shape(sGt)[0]):
        v += f"{sGt[k, :]};"
    v += "]"
    v = v.replace("\n", "")
    v = re.sub(" +", " ", v)  # remove extra white spaces
    # v = re.sub(". ", " ", v)  # remove point after integers

    formatted_sGt = v

    return formatted_sGt


def multifit_apply_noisegate(
    gain_table: ndarray,
    sFit_model_frequencies: ndarray,
    sFit_model_levels: ndarray,
    noisegate_level: ndarray,
    noisegate_slope: ndarray,
) -> ndarray:
    """Apply noisegate to the gain table

    Based on OpenMHA subfunction of libmultifit.m.

    Args:
        gain_table (ndarray): gain array
        sFit_model_frequencies (list): FFT filterbank frequencies
        sFit_model_levels (ndarray): levels at which to calculate gains
        noisegate_level (ndarray): chosen compression threshold
        noisegate_slope (ndarray): determines slope below compression threshold

    Returns:
        ndarray: gain table with noisegate applied

    """
    gain_table_corrected = gain_table.copy()
    for i in [0, 1]:
        for kf in range(len(sFit_model_frequencies)):
            gain_noisegate = interp1d(
                sFit_model_levels,
                gain_table_corrected[:, kf, i],
                fill_value="extrapolate",
            )(noisegate_level[kf, i])
            idx = [
                i for i, x in enumerate(sFit_model_levels < noisegate_level[kf, i]) if x
            ]
            gain_table_corrected[idx, kf, i] = (
                sFit_model_levels[idx] - noisegate_level[kf, i]
            ) * noisegate_slope[kf, i] + gain_noisegate

    return gain_table_corrected
