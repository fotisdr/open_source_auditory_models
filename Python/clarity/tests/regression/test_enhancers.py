import json
import tempfile

import numpy as np
import torch

from clarity.enhancer.dsp import filter  # pylint: disable=redefined-builtin
from clarity.enhancer.gha.gha_interface import GHAHearingAid as gha
from clarity.enhancer.gha.gha_utils import format_gaintable, get_gaintable
from clarity.utils.audiogram import Audiogram
from clarity.utils.file_io import read_signal

gha_params = {  # hyperparameters for GHA Hearing Aid, BE CAREFUL if making changes
    "sample_rate": 44100,
    "ahr": 20,
    "audf": None,
    "cfg_file": "prerelease_combination3_smooth",
    "noise_gate_levels": None,
    "noise_gate_slope": 0,
    "cr_level": 0,
    "max_output_level": 100,
    "equiv_0db_spl": 100,
    "test_nbits": 16,
}


def test_dsp_filter(regtest):
    amfir = filter.AudiometricFIR(sample_rate=44100, nfir=220, device="cpu")
    torch.manual_seed(0)
    signal = torch.rand(10, dtype=torch.float)
    signal = torch.reshape(signal, (1, 1, -1))
    output = amfir(signal.to(amfir.device))
    output = np.round(output.detach().cpu().numpy(), 4)
    regtest.write(f"signal output: \n{output}\n")


def test_gha_audiogram(regtest):
    frequencies = np.array([250, 500, 1000, 2000, 4000, 8000])
    for i in [1, 10, 30, 40]:
        np.random.seed(0)
        levels_l = np.round(np.log10(np.random.rand(6) / 20) * (-i), 0)
        levels_r = np.round(np.log10(np.random.rand(6) / 20) * (-i), 0)
        ag_left = Audiogram(levels=levels_l, frequencies=frequencies)
        ag_right = Audiogram(levels=levels_r, frequencies=frequencies)
        regtest.write(
            "Audiogram original: \n"
            f"{ag_left.frequencies}\n{ag_left.levels}\n{ag_right.levels}\n"
            f"['{ag_left.severity}', '{ag_right.severity}']\n"
        )
        ag_left = ag_left.resample(np.array([500, 1000, 2000]))
        ag_right = ag_right.resample(np.array([500, 1000, 2000]))
        regtest.write(
            f"Audiogram new: \n{ag_left.frequencies}\n{ag_left.levels}\n"
            f"{ag_right.levels}\n['{ag_left.severity}', '{ag_right.severity}']\n"
        )


def test_GHA_inputs(regtest):
    enhancer = gha(**gha_params)

    infile_names = [
        f"tests/test_data/scenes/S06001_mixed_CH{ch}.wav" for ch in range(1, 4)
    ]
    _fd_merged, merged_filename = tempfile.mkstemp(
        prefix="clarity-merged-", suffix=".wav"
    )
    enhancer.create_HA_inputs(infile_names, merged_filename)

    signal = read_signal(merged_filename)
    np.set_printoptions(threshold=100)

    # outputting a segment not at start to avoid it being all zeros
    regtest.write(f"signal output: \n{signal[1000:1010, :]}\n")


def test_GHA_config(regtest):
    np.set_printoptions(threshold=1000)
    enhancer = gha(**gha_params)
    with open("tests/test_data/metadata/listeners.json", encoding="utf-8") as fp:
        listeners = json.load(fp)
    listener = listeners["L0001"]
    frequencies = np.array(listener["audiogram_cfs"], dtype="int")
    audiogram_left_levels = np.array(listener["audiogram_levels_l"])
    audiogram_right_levels = np.array(listener["audiogram_levels_r"])
    audiogram_left = Audiogram(frequencies=frequencies, levels=audiogram_left_levels)
    audiogram_right = Audiogram(frequencies=frequencies, levels=audiogram_right_levels)

    cfg_template = f"tests/test_data/openMHA/{enhancer.cfg_file}_template.cfg"

    gaintable = get_gaintable(
        audiogram_left,
        audiogram_right,
        enhancer.noise_gate_levels,
        enhancer.noise_gate_slope,
        enhancer.cr_level,
        enhancer.max_output_level,
    )
    formatted_sGt = format_gaintable(gaintable, noisegate_corr=True)

    config = enhancer.create_configured_cfgfile(
        "input", "output", formatted_sGt, cfg_template
    )
    regtest.write(f"config output: \n{config}\n")
