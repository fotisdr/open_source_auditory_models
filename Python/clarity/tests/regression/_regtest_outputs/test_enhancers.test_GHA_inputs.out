signal output:
[[-0.30786133 -0.39199829 -0.34420776 -0.35498047]
 [-0.30331421 -0.4085083  -0.33901978 -0.37487793]
 [-0.29840088 -0.42907715 -0.33230591 -0.40112305]
 [-0.29516602 -0.4453125  -0.32678223 -0.41842651]
 [-0.29556274 -0.45758057 -0.32110596 -0.43707275]
 [-0.29144287 -0.47451782 -0.31506348 -0.46224976]
 [-0.28607178 -0.48565674 -0.31271362 -0.48049927]
 [-0.28259277 -0.49224854 -0.31155396 -0.49224854]
 [-0.28088379 -0.49679565 -0.30471802 -0.50222778]
 [-0.28219604 -0.49954224 -0.29943848 -0.51016235]]
